pkg/receive/hashring.go

```diff
@@ -0,0 +1,74 @@
+package receive
+
+import (
+	"errors"
+	"sort"
+
+	"github.com/improbable-eng/thanos/pkg/store/prompb"
+
+	"github.com/cespare/xxhash"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const sep = '\xff'
+
+// Hashring finds the correct host to handle a given time series
+// for a specified tenant.
+// It returns the hostname and any error encountered.
+type Hashring interface {
+	GetHost(string, *prompb.TimeSeries) (string, error)
+}
+
+// hash returns a hash for the given tenant and time series.
+func hash(tenant string, ts *prompb.TimeSeries) uint64 {
+	// Sort labelset to ensure a stable hash.
+	sort.Slice(ts.Labels, func(i, j int) bool { return ts.Labels[i].Name < ts.Labels[j].Name })
+
+	b := make([]byte, 0, 1024)
+	b = append(b, []byte(tenant)...)
+	for _, v := range ts.Labels {
+		b = append(b, v.Name...)
+		b = append(b, sep)
+		b = append(b, v.Value...)
+		b = append(b, sep)
+	}
+	return xxhash.Sum64(b)
+}
+
+// simpleHashring represents a group of hosts handling write requests.
+type simpleHashring []string
+
+// GetHost returns a hostname to handle the given tenant and time series.
+func (s simpleHashring) GetHost(tenant string, ts *prompb.TimeSeries) (string, error) {
+	// Always return nil here to implement the Hashring interface.
+	return s[hash(tenant, ts)%uint64(len(s))], nil
+}
+
+// multiTenantHashring represents a set of Hashrings for different tenants.
+type multiTenantHashring map[string]Hashring
+
+// GetHost returns a hostname to handle the given tenant and time series.
+func (m multiTenantHashring) GetHost(tenant string, ts *prompb.TimeSeries) (string, error) {
+	if h, ok := m[tenant]; ok {
+		return h.GetHost(tenant, ts)
+	}
+	if h, ok := m[""]; ok {
+		return h.GetHost(tenant, ts)
+	}
+	return "", errors.New("no matching hosts to handle tenant")
+}
+
+// NewHashring creates a multi-tenant hashring for a given
+// configuration that maps tenants to groups of hosts.
+func NewHashring(groups []*targetgroup.Group) Hashring {
```

Comment:
This can technically change at runtime, no? I'm ok with skipping this aspect for now, but we either need to make clear that only what is discovered at startup is what is discovered over the lifetime of the process.

Reply:
I actually wrote another hashring implementation called `discoveryHashring`, which updates the hashring as new peers are discovered, but decided to throw it out as we need need to shut down the receiver *anyways* when new peers are discovered, so there was little benefit in having a dynamic hashring right now.

The other choice is to make it updateable on the hashring-side of things and still cleanly shutdown the receiver anyways when the hashring changes. WDYT?

Reply:
Yeah, I was more saying that if we ever detect a change of these groups at runtime, we should log that that has happened.